datastore/gcloud/rest/datastore/datastore.py

```diff
@@ -0,0 +1,412 @@
+
+        if IS_DEV:
+            self._project = os.environ.get('DATASTORE_PROJECT_ID', 'dev')
+            # Tokens are not needed when using dev emulator
+            self.token = None
+        else:
+            self._project = project
+            self.token = token or Token(service_file=service_file,
+                                        session=session, scopes=SCOPES)
+
+    def project(self):
+        # type: () -> str
+        if self._project:
+            return self._project
+
+        self._project = self.token.get_project()
+        if self._project:
+            return self._project
+
+        raise Exception('could not determine project, please set it manually')
+
+    @staticmethod
+    def _make_commit_body(mutations, transaction=None,
+                          mode=Mode.TRANSACTIONAL):
+        # type: (List[Dict[str, Any]], Optional[str], Mode) -> Dict[str, Any]
+        if not mutations:
+            raise Exception('at least one mutation record is required')
+
+        if transaction is None and mode != Mode.NON_TRANSACTIONAL:
+            raise Exception('a transaction ID must be provided when mode is '
+                            'transactional')
+
+        data = {
+            'mode': mode.value,
+            'mutations': mutations,
+        }
+        if transaction is not None:
+            data['transaction'] = transaction
+        return data
+
+    def headers(self):
+        # type: () -> Dict[str, str]
+        if IS_DEV:
+            return {}
+
+        token = self.token.get()
+        return {
+            'Authorization': 'Bearer %s' % token,
+        }
+
+    # TODO: support mutations w version specifiers, return new version (commit)
+    @classmethod
+    def make_mutation(cls, operation, key, properties=None):
+        # type: (Operation, Key, Dict[str, Any]) -> Dict[str, Any]
+        if operation == Operation.DELETE:
+            return {operation.value: key.to_repr()}
+
+        return {
+            operation.value: {
+                'key': key.to_repr(),
+                'properties': {k: cls.value_kind(v).to_repr()
+                               for k, v in properties.items()},
+            }
+        }
+
+    # https://cloud.google.com/datastore/docs/reference/data/rest/v1/projects/allocateIds
+    def allocateIds(self, keys, session=None, timeout=10):
+        # type: (List[Key], requests.Session, int) -> List[Key]
+
+        project = self.project()
+        url = '%s/projects/%s:allocateIds' % (API_ROOT, project)
+
+        payload = json.dumps({
+            'keys': [k.to_repr() for k in keys],
+        }).encode('utf-8')
+
+        headers = self.headers()
+        headers.update({
+            'Content-Length': str(len(payload)),
+            'Content-Type': 'application/json',
+        })
+
+        if not self.session:
+            self.session = requests.Session()
+        session = session or self.session
+        resp = session.post(url, data=payload,
```

Comment:
All of these callouts will need to be wrapped in a lock, as with the other gcloud-rest-* projects. Similarly, we'll need to add that argument to `Datastore.__init__()`